econml/cate_estimator.py

```diff
@@ -456,6 +460,93 @@ def const_marginal_effect_inference(self, X=None):
+            The name of the outcome.
+
+        Returns
+        -------
+        shap_outs: nested dictionary of Explanation object
+            A nested dictionary by using each Y and each T as key and the shap_values explanation object as value.
+
+
+        """
+        d_t = self._d_t[0] if self._d_t else 1
+        d_y = self._d_y[0] if self._d_y else 1
+        if treatment_names is None:
+            treatment_names = [f"T{i}" for i in range(d_t)]
+        if output_names is None:
+            output_names = [f"Y{i}" for i in range(d_y)]
+
+        return _shap_explain_cme(self.const_marginal_effect, X, d_y, d_t, feature_names, treatment_names, output_names)
+
+    def _shap_values(self, models, X, *, feature_names=None, treatment_names=None, output_names=None):
+        """ Shap value for the final stage models.
+        This is the private method mainly for internal use.
+
+        Parameters
+        ----------
+        models: a single estimator or a list of estimators with one estimator per treatment
+            models for the model's final stage model.
+        X: (m, d_x) matrix
+            Features for each sample. Should be in the same shape of fitted X in final stage.
+        feature_names: optional None or list of strings of length X.shape[1] (Default=None)
+            The names of input features.
+        treatment_names: optional None or list (Default=None)
+            The name of treatment. In discrete treatment scenario, the name should not include control name.
+        output_names:  optional None or list (Default=None)
+            The name of the outcome.
+
+        Returns
+        -------
+        shap_outs: nested dictionary of Explanation object
+            A nested dictionary by using each Y and each T as key and the shap_values explanation object as value.
+        """
+
+        d_t = self._d_t[0] if self._d_t else 1
+        d_y = self._d_y[0] if self._d_y else 1
+        if treatment_names is None:
+            treatment_names = [f"T{i}" for i in range(d_t)]
+        if output_names is None:
+            output_names = [f"Y{i}" for i in range(d_y)]
+
+        if not isinstance(models, list):
+            models = [models]
+        assert len(models) == d_t, "Number of final stage models don't equals to number of treatments!"
+        # define masker by using entire dataset, otherwise Explainer will only sample 100 obs by default.
+        background = shap.maskers.Independent(X, max_samples=X.shape[0])
+
+        shap_outs = defaultdict(dict)
+        for i in range(d_t):
+            try:
+                explainer = shap.Explainer(models[i], background,
+                                           feature_names=feature_names)
+            except Exception as e:
+                print("Final model can't be parsed, explain const_marginal_effect() instead!")
+                return _shap_explain_cme(self.const_marginal_effect, X, d_y, d_t, feature_names, treatment_names,
```

Comment:
What happens if only one of the models fails to be built by the explainer? Why not use _shap_explain_cme only for that coordinate??

We could internally to _shap_explain_cme, build a wrapper for const_marginal_effect that returns only the i-th coordinate of const_marginal_effect and pass that function to shap. Then shap will return shap_Values only for that coordinate and we could be using the model explainers for the rest. 

Though I guess it's not clear what the compiutational benefit will be with this, since once you call shap on a generic function it doesn't change much the computation if that function is single output or multi-output. So maybe the current version is good too.

Reply:
Is this addressed?